File: java/src/main/java/falcon/api/Permissions.java
package falcon.api;

import falcon.api.internal.Interop;
import falcon.api.internal.PluginContext;

import java.lang.foreign.Arena;

public final class Permissions {
    private final PluginContext mContext;

    Permissions(PluginContext context) {
        mContext = context;
    }

    public boolean add(String node) {
        return add(node, PermissionDefault.OPERATOR);
    }

    public boolean add(String node, PermissionDefault defaultValue) {
        try (Arena arena = Arena.ofConfined()) {
            return Interop.api().registerPermission(mContext.handle(), Interop.text(arena, node),
                    defaultValue.value()) != 0;
        }
    }
}

File: dotnet/Falcon.PluginAPI/Permissions/Permissions.cs
using Falcon.Interop;

namespace Falcon
{
    public sealed unsafe class Permissions
    {
        private readonly nint _plugin;

        internal Permissions(nint plugin)
        {
            _plugin = plugin;
        }

        public bool Add(string node, PermissionDefault defaultValue = PermissionDefault.Operator)
        {
            fixed (byte* text = NativeApi.Utf8(node))
            {
                return NativeApi.Table->registerPermission((FalconPlugin*)_plugin, text, (uint)defaultValue) != 0;
            }
        }
    }
}
